File: LP2/AI/assignment5.cpp
#include <iostream>
#include <string>
using namespace std;

class ChatBot {
public:
    void startChat() {
        cout << "Hello! I am your simple ChatBot.\n";
        cout << "You can ask me about: weather, time, hobby, or type 'exit' to leave.\n";
        string input;
        
        while (true) {
            cout << "\nYou: ";
            getline(cin, input);
            toLowerCase(input);
            
            if (input == "exit") {
                cout << "ChatBot: Goodbye! Have a nice day!\n";
                break;
            } 
            else {
                respond(input);
            }
        }
    }

private:
    void respond(string question) {
        if (question.find("weather") != string::npos) {
            cout << "ChatBot: The weather is sunny and bright!";
        } 
        else if (question.find("time") != string::npos) {
            cout << "ChatBot: It's chatbot time — always ready for you!";
        } 
        else if (question.find("hobby") != string::npos) {
            cout << "ChatBot: I love chatting with smart people like you!";
        } 
        else {
            cout << "ChatBot: Sorry, I don't understand that.";
        }
    }
    
    void toLowerCase(string &str) {
        for (int i = 0; i < str.length(); i++) {
            str[i] = tolower(str[i]);
        }
    }
};

int main() {
    ChatBot bot;
    bot.startChat();
    return 0;
}

File: LP2/AI/assignment5.py
from sentence_transformers import SentenceTransformer, util #type: ignore
import torch #type: ignore

faq_pairs = {
    "What is your name?": "I'm a customer support assistant.",
    "Do you offer refunds?": "Yes, we offer refunds within 30 days of purchase.",
    "Tell me about the product.": "Can you specify which product you're interested in?",
    "What is the price of the product?": "Please mention the product name for pricing details.",
    "Hello": "Hi there! How can I assist you today?",
    "Thank you": "You're welcome!",
    "Bye": "Goodbye! Have a great day.",
}


# Load pre-trained model
model = SentenceTransformer('paraphrase-MiniLM-L6-v2')

# Prepare data
questions = list(faq_pairs.keys())
question_embeddings = model.encode(questions, convert_to_tensor=True)

def get_response(user_input):
    user_embedding = model.encode(user_input, convert_to_tensor=True)
    cosine_scores = util.pytorch_cos_sim(user_embedding, question_embeddings)
    top_result = torch.argmax(cosine_scores)

    if cosine_scores[0][top_result] > 0.6:
        return faq_pairs[questions[top_result]]
    else:
        return "I'm sorry, I didn't understand that. Could you please rephrase?"


import streamlit as st #type: ignore

st.title("🧠 NLP Chatbot (BERT-powered)")
st.write("Ask a question about our services!")

user_input = st.text_input("You:", "")

if user_input:
    response = get_response(user_input)
    st.text_area("Bot:", response, height=100)